content/upcoming/enry.md

```diff
@@ -0,0 +1,206 @@
+
+## Why develop a port?
+
+Some time ago, at source{d}, we needed to detect the language of every file in every repository to perform analysing job in our previous pipeline. To face this task, we looked at [github's linguist project](https://github.com/github/linguist), a Ruby library that offered what we needed, but didn't address all our requirements though.
+
+So at this point, we realised we needed to implement an approach that fit in an environment mostly developed in Go with a strong performance and that kept its compatibility with the mother project. Which is why *enry* was born.
+
+At the beginning, *enry* had only a sub-set of the *linguist* functionality, enough to cover basic needs of the analysis pipeline that we had back then.
+
+As time passed, *enry* showed it was a  valuable tool that could be used in other projects (e.g.[babelfish](https://github.com/bblfsh)) and it started to incorporate almost all the features *linguist* has.
+
+## How does enry work?
+
+Detecting a file language is done by the functions:
+
+```go
+func GetLanguage(filename string, content []byte) (language string)
+func GetLanguages(filename string, content []byte) []string
+```
+
+After calling `GetLanguages`, there could be more than one possible language detected for the file, so it returns all of them, and `GetLanguage` returns one of them (in fact, the most likely one thanks to the classifier - which we'll cover later).
+
+The detection process must be split into several steps that form a chain or sequence of strategies. These strategies need the file name, its content and a list of candidates (a list of possible languages for the file). Strategies are typed functions with the next signature:
+
+```go
+type Strategy func(filename string, content []byte, candidates []string) (languages []string)
+```
+
+Strategies try to guess the language of the file based on these arguments and a specific characteristic. For example, the default strategies enry uses are:
+
+```go
+var DefaultStrategies = []Strategy{
+    GetLanguagesByModeline,
+    GetLanguagesByFilename,
+    GetLanguagesByShebang,
+    GetLanguagesByExtension,
+    GetLanguagesByContent,
+    GetLanguagesByClassifier,
+}
+```
+
+You can see how the strategies look for `Modeline`, `Filename`, `Shebang`... characteristics of a file that can be representative of the language it uses. Some strategies need to parse the content and perhaps apply heuristics while others do their job only with the filename.
+
+The strategy chain works as follows:
+* A strategy tries to get the language from a file.
+* It can result in zero or more languages detected.
+* If there's no language, call the next strategy.
+* If there are more than one language, they are appended to a list of candidates and they are given to the next strategy.
```

Comment:
there is

https://english.stackexchange.com/questions/35389/there-is-are-more-than-one-whats-the-difference